runtime/src/main/java/org/corfudb/runtime/collections/LRMultiNamespaceListener.java

```diff
@@ -0,0 +1,86 @@
+package org.corfudb.runtime.collections;
+
+import lombok.Setter;
+import org.corfudb.runtime.LogReplication.ReplicationStatusVal;
+import org.corfudb.runtime.view.ObjectsView;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This is the callback interface that any client subscribing to updates across the LR Status Table
+ * (ReplicationStatus) and application-specific tables must implement.
+ *
+ * This interface sees ordered updates from :
+ * 1. client-streams from client-Namespace, and,
+ * 2. LrStatusTable from corfuSystem-Namespace.
+ */
+public abstract class LRMultiNamespaceListener implements StreamListener {
+
+    private boolean snapshotSyncInProgress = false;
+
+    private static final String REPLICATION_STATUS_TABLE = ObjectsView.REPLICATION_STATUS_TABLE;
+
+    @Setter
+    long subscriptionTs = 0;
+
+    public void onNextEntry(CorfuStreamEntries results) {
+        onNext(results);
+    }
+
+    /**
+     * A corfu update can/may have multiple updates belonging to different streams.
+     * This callback will return those updates as a list grouped by their Stream UUIDs.
+     *
+     * Note: there is no order guarantee within the transaction boundaries.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    public void onNext(CorfuStreamEntries results) {
+
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+
+        for (TableSchema tableSchema : entries.keySet()) {
+            if (Objects.equals(tableSchema.getTableName(), REPLICATION_STATUS_TABLE)) {
```

Comment:
Actually session is hard to get from client, probably subscriber, and we'll have to query all sessions between any source and us, with thee given subscriber. So that covers the case of sharding?

Reply:
Yes, but it cannot be done in this PR as it's tracking master.